doc/libsecrm.3
.\"	Process this file with groff -man -Tascii foo.3
.\"
.TH LibSecRm 3 GNU/Linux "User's Manual"

.SH NAME
LibSecRm \- Ensuring secure removal of data.

.SH SYNOPSIS
~/.bash_profile:

.B export LD_PRELOAD=/usr/local/lib/libsecrm.so

and/or

/etc/ld.so.preload:

.B /usr/local/lib/libsecrm.so

.SH DESCRIPTION
The LibSecRm library ensures that data removed from files is removed securely
(just like using the 'shred' program), by intercepting necessary library calls:
unlink, unlinkat, remove (obvious), fopen, freopen, truncate, ftruncate,
open, openat (with O_TRUNC passed in flags), creat, and the 64-bit versions of all these. The library
.B MUST NOT
be setuid(root). This would allow anyone to delete files from system directories.

LibSecRm partially also takes care about memory security - some of the memory
allocation functions are intercepted (namely: malloc, posix_memalign, valloc, memalign,
brk, sbrk), and the allocated memory is wiped before
passing it to the calling program.

Selected programs can be taken out of LibSecRm's control by being put in banning files.
A banning files consists of program or file names, one per line. If a text found in a line
of a banning file is found in the program's name, LibSecRm does
.B NOT
wipe the data it would wipe otherwise.

.SH NOTES
The file and command given in the synopsis section may be different on shells other than BASH.

.SH FILES
The /proc filesystem is used to check if a file to be wiped is already opened.

${sysconfdir} is /usr/local/etc unless defined otherwise.

${sysconfdir}/libsecrm.progban - which programs want to run normally (partial names are enough). The
/proc filesystem must be mounted in order for this to work right now.

${sysconfdir}/libsecrm.fileban - which files not to wipe (partial names are enough)

$HOME/.libsecrm.progban - path to additional user program banning file

$HOME/.libsecrm.fileban - path to additional user file banning file

Put each entry on its own line.

.SH ENVIRONMENT
LIBSECRM_PROGBANFILE - path to an additional program banning file

LIBSECRM_FILEBANFILE - path to an additional file banning file

.SH AUTHOR
Bogdan 'bogdro' Drozdowski,
bogdandr /AT/ op . pl

bogdro /AT/ rudy . mif . pg . gda . pl

[EN] http://libsecrm.sf.net

[PL] http://libsecrm.sf.net/index-pl.php

[EN] http://bogdro.evai.pl/soft/

[EN] http://bogdro.evai.pl/inne/

[PL] http://bogdro.evai.pl/

[EN] http://bogdro.evai.pl/index-en.php

[EN] http://bogdro.byethost18.com/soft/

[EN] http://bogdro.byethost18.com/inne/

[PL] http://bogdro.byethost18.com/

[EN] http://bogdro.byethost18.com/index-en.php

[EN] http://rudy.mif.pg.gda.pl/~bogdro/soft/

[EN] http://rudy.mif.pg.gda.pl/~bogdro/inne/

[PL] http://rudy.mif.pg.gda.pl/~bogdro/

.SH THANKS TO

  - Manuel Arriaga for libtrash, parts of which are used here
  - Colin Plumb, for the great 'shred' program, parts of which are used here.
    The 'shred' utility is:
 	Copyright (C) 1999-2006 Free Software Foundation, Inc.
 	Copyright (C) 1997, 1998, 1999 Colin Plumb.
  - The authors of Libsafe for the great piece of software, parts of which
    are used here. Libsafe is:
 	Copyright (C) 2002 Avaya Labs, Avaya Inc.
 	Copyright (C) 1999 Bell Labs, Lucent Technologies.
 	Copyright (C) Arash Baratloo, Timothy Tsai, and Navjot Singh.
  - The authors of the "fuser" utility, parts of which are used here. Fuser is:
 	Based on fuser.c Copyright (C) 1993-2005 Werner Almesberger and Craig Small
 	Completely re-written
 	Copyright (C) 2005 Craig Small


.SH "SEE ALSO"
.BR shred (1),
.B info wipefreespace

The 'info' documentation is more descriptive and is the main one. Type

.B 	info libsecrm

to read it.



ChangeLog
2021-01-09	Bogdan Drozdowski <bogdro@users.sourceforge.net>

	* LibSecRm version 2.9
	  Added intercepting posix_fallocate64(). Portability improvements.
	  LibSecRm should work better with some filesystems which have i-node
	  numbers greater than 2^32-1 and compile under non-ANSI-C. Checked
	  running LibSecRm under GCC address & undefined behaviour sanitizers.
	  Added more unit tests.
	* configure.ac: added checking for the stat64(), lstat64() and 
	  fstatat64() functions. Updated the test for the ptrdiff_t type by
	  adding stddef.h (and a check for it). Added creating doc/libsecrm.3.
	  Checking for the mkfifo() function for future unit tests. Checking
	  for the posix_fallocate64() function. Changed the 'intercept-malloc'
	  option to be '--enable-', not '--with-'. Added checking for the
	  -Wstringop-truncation and -Wduplicated-branches compiler options.
	* m4: removed unused files
	* src/lsr_cfg.h.in: added new constants
	* src/*.c*: added a preprocessor block that marks ANSI C as disabled,
	  for internal testing compiler compatibility.
	* src/lsr_unlink.c (__lsr_rename): choose a new character to wipe the
	  filename if the chosen one is the same as the original (at least the
	  first one).
	* src/lsr_unlink.c (rmdir): prefer lstat64() if available. Check also
	  for lstat() and skip if not available.
	* src/lsr_unlink.c (unlinkat): open the file in exclusive mode, just
	  like in the other functions
	* src/lsr_banning.c (check_dir, __lsr_is_forbidden_fs,
	  __lsr_can_wipe_filename): prefer stat64() if available. Check
	  also for stat() and skip if not available.
	* src/lsr_banning.c (__lsr_can_wipe_dirname, __lsr_is_forbidden_file):
	  prefer lstat64() if available. Check also for lstat() and skip if
	  not available.
	* src/lsr_banning.c (__lsr_can_wipe_filename_atdir): prefer fstatat64()
	  if available. Check also for fstatat() and skip if not available.
	* src/lsr_banning.c (__lsr_can_wipe_filedesc): prefer fstat64()
	  if available. Check also for fstat() and skip if not available.
	* src/lsr_banning.c (__lsr_is_forbidden_fd): fix a potential overflow
	* src/lsr_banning.c (__lsr_is_forbidden_file): replace rindex() with
	  the more portable strrchr().
	* src/lsr_wiping.c (__lsr_fd_truncate): prefer fstat64() if available.
	  Check also for fstat() and skip if not available.
	* src/lsr_truncate.c (posix_fallocate64): a new intercepted function
	* src/lsr_truncate.c (generic_posix_fallocate): a new generic function
	  to support both posix_fallocate() and posix_fallocate64().
	* src/lsr_truncate.c (fallocate, generic_posix_fallocate): Check also
	  for fstat() and skip if both fstat64() and fstat() are not available.
	* src/lsr_opens.c (generic_fopen, generic_freopen, generic_open,
	  generic_openat): code simplification and size reduce
	* src/lsr_public.c.in (lsr_posix_fallocate, lsr_fallocate): fixed
	  compiling without ANSI C
	* src/lsr_priv.h.in: corrected some declarations to avoid warnings
	* src/banning_generic.c (__banning_is_banned): fixed using the user
	  banning filename instead of the global one
	* Makefile.am: stop using 'tar --delete', for systems where GNU tar is
	  not installed or is not the default. Marked the 'pack' target PHONY
	* test/*.c, test/lsrtest_common.h: added a prolog macro with logging
	  common for all tests
	* test/lsrtest_banning.c, test/lsrtest_unlink.c: moved common code to
	  separate functions.
	* test/*.c: Fixed some compiler warnings.
	* test/lsrtest_unlink.c: free()ing pointers in tests when GNU libc is
	  detected (it's allowed in this case and removes sanitizer errors).
	  Moved test_unlink_banned() out of the unlinkat() test block. Added
	  tests for unlinkat() and remove() with symbolic links.
	* test/lsrtest_fopens.c: added tests for a device file and for an
	  object in the /proc filesystem. Added tests for the "w+" open mode.
	* test/lsrtest_other.c: a new test for finding a private symbol in
	  the library and for the function filling the wiping buffer.
	* test/lsrtest_truncate.c: added a test for posix_fallocate64()
	* test/Makefile.am: added the new lsrtest_other test, added a target
	  to just compile the tests
	* doc/libsecrm.3.in: created from doc/libsecrm.3 to allow ./configure
	  substitutions later.
	* doc/Makefile.am: added distributing libsecrm.3.in
	* doc/libsecrm.texi.in, doc/libsecrm.3.in: updated the addresses
	* doc/libsecrm.texi.in: more blocks in @command{}. Added the new
	  public function - lsr_posix_fallocate64(). Substituting the return
	  and argument types for lsr_brk() and lsr_sbrk()
	* libsecrm.spec.in: removed obsolete commented-out commands, updated
	  the URL, added BugURL. Using macros for common elements.


2019-02-08	Bogdan Drozdowski <bogdro@users.sourceforge.net>

	* LibSecRm version 2.8
	  Added intercepting new functions. Fixed initialization code - fixed
	  lookup for fopen() and made intercepting malloc() disabled by
	  default, because it causes the library to crash. Many improvements
	  in checking if an object is banned from being wiped. Improvements
	  in code portability and compatibility. Better code maintainability,
	  updated copyright and documentation. Improvements in unit tests and
	  new tests added.
	* configure.ac: moved AC_LANG(C) before AC_PROG_LIBTOOL for improved
	  compatibility. Checking for sys/sysmacros.h for compatibility with
	  future glibc versions, which define makedev() there. Added checking
	  for the -Wno-nonnull-compare compiler flag to avoid warnings about
	  defensive programming code. Added checking for the stat() function.
	  Added checking for the -Wchkp, -Wformat-overflow=2, -Wrestrict
	  -Wduplicated-cond, and -Woverlength-strings compiler warning options
	* missing: script updated from new autoconf
	* src/libsecrm.c (__lsr_copy_string): put variables before code
	* src/libsecrm.c (__lsr_main): better casting of time() result to
	  avoid warnings. Set the internal flag when initializing. When the

src/libsecrm.c
/*
 * A library for secure removing data.
 *
 * Copyright (C) 2007-2021 Bogdan Drozdowski, bogdro (at) users . sourceforge . net
 * License: GNU General Public License, v3+
 *
 * Syntax example: export LD_PRELOAD=/usr/local/lib/libsecrm.so
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foudation:
 *		Free Software Foundation
 *		51 Franklin Street, Fifth Floor
 *		Boston, MA 02110-1301
 *		USA
 *
 * Thanks to:
 * - Manuel Arriaga for libtrash, parts of which are used here
 * - Colin Plumb, for the great 'shred' program, parts of which are used here.
 *   The 'shred' utility is:
 *	Copyright (C) 1999-2006 Free Software Foundation, Inc.
 *	Copyright (C) 1997, 1998, 1999 Colin Plumb.
 * - The authors of Libsafe for the great piece of software, parts of which
 *   are used here. Libsafe is:
 *	Copyright (C) 2002 Avaya Labs, Avaya Inc.
 *	Copyright (C) 1999 Bell Labs, Lucent Technologies.
 *	Copyright (C) Arash Baratloo, Timothy Tsai, and Navjot Singh.
 * - The authors of the "fuser" utility, parts of which are used here. Fuser is:
 *	Based on fuser.c Copyright (C) 1993-2005 Werner Almesberger and Craig Small
 *	Completely re-written
 *	Copyright (C) 2005 Craig Small
 *
 */

#include "lsr_cfg.h"

#if (defined HAVE_DLFCN_H) && ((defined HAVE_DLSYM) || (defined HAVE_LIBDL))
	/* need RTLD_NEXT and dlvsym(), so define _GNU_SOURCE */
# ifndef _GNU_SOURCE
#  define _GNU_SOURCE	1
# endif
# include <dlfcn.h>
# ifndef RTLD_NEXT
#  define RTLD_NEXT ((void *) -1l)
# endif
#else
# ifdef LSR_ANSIC
#  error Dynamic loading functions missing.
# endif
#endif

#include <stdio.h>

	/* time() for randomization purposes */
#ifdef TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# ifdef HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  ifdef HAVE_TIME_H
#   include <time.h>
#  endif
# endif
#endif

#ifdef HAVE_STDLIB_H
# include <stdlib.h>	/* random(), srandom(), rand(), srand() */
#endif

#ifdef HAVE_STRING_H
# if (!defined STDC_HEADERS) && (defined HAVE_MEMORY_H)
#  include <memory.h>
# endif
# include <string.h>
#endif

#include "lsr_priv.h"

static int	__lsr_is_initialized	= LSR_INIT_STAGE_NOT_INITIALIZED;

/* Pointers to original functions */
static i_cp		__lsr_real_unlink		= NULL;
static i_cp		__lsr_real_remove		= NULL;
static i_i_cp_i		__lsr_real_unlinkat		= NULL;
static i_cp		__lsr_real_rmdir		= NULL;

static i_cp_o64		__lsr_real_truncate64		= NULL;
static i_i_o64		__lsr_real_ftruncate64		= NULL;
static fp_cp_cp		__lsr_real_fopen64		= NULL;
static fp_cp_cp_fp	__lsr_real_freopen64		= NULL;
static i_cp_i_		__lsr_real_open64		= NULL;
static i_i_cp_i_	__lsr_real_openat64		= NULL;
static i_cp_mt		__lsr_real_creat64		= NULL;

static i_cp_o		__lsr_real_truncate		= NULL;
static i_i_o		__lsr_real_ftruncate		= NULL;
static fp_cp_cp		__lsr_real_fopen		= NULL;
static fp_cp_cp_fp	__lsr_real_freopen		= NULL;
static i_cp_i_		__lsr_real_open			= NULL;
static i_i_cp_i_	__lsr_real_openat		= NULL;
static i_cp_mt		__lsr_real_creat		= NULL;
static i_o_o		__lsr_real_posix_fallocate	= NULL;
static i_o64_o64	__lsr_real_posix_fallocate64	= NULL;
static i_i_i_o_o	__lsr_real_fallocate		= NULL;

/* memory-related functions: */
static f_s		__lsr_real_malloc		= NULL;
static vpp_s_s		__lsr_real_psx_memalign		= NULL;
static f_s		__lsr_real_valloc		= NULL;
static f_s		__lsr_real_pvalloc		= NULL;
static f_s_s		__lsr_real_memalign		= NULL;
static f_s_s		__lsr_real_aligned_alloc	= NULL;
static f_vp		__lsr_real_brk			= NULL;
static f_ip		__lsr_real_sbrk			= NULL;

#ifdef TEST_COMPILE
# undef LSR_ANSIC
#endif

/* =============================================================== */

#if (defined __STRICT_ANSI__) || (!defined HAVE_SRANDOM) || (!defined HAVE_RANDOM)

static unsigned long int __lsr_next = 0xdeafface;

/* 'man rand': */
int __lsr_rand (LSR_VOID)
{
	__lsr_next = __lsr_next * 1103515245 + 12345;
	return ((unsigned int)(__lsr_next/65536) % 32768);
}

static void __lsr_srand (
# ifdef LSR_ANSIC
	unsigned int seed)
# else
	seed)
	unsigned int seed;
# endif
{
	__lsr_next = seed;
}
#endif

/* =============================================================== */

void __lsr_copy_string (
#ifdef LSR_ANSIC
	char * const dest, const char src[], const size_t len)
#else
	dest, src, len)
	char * const dest;
	const char src[];
	const size_t len;
#endif
{
#if (!defined HAVE_STRING_H) && (!defined HAVE_MEMCPY)
	size_t i;
#endif
	if ( (src == NULL) || (dest == NULL) )
	{
		return;
	}
#ifdef HAVE_STRING_H
	strncpy (dest, src, len);
#else
# if defined HAVE_MEMCPY
	memcpy (dest, src, len);
# else
	for ( i = 0; i < len; i++ )
	{
		dest[i] = src[i];
	}
# endif
#endif
	dest[len] = '\0';
}

/* =============================================================== */

#ifndef HAVE_MEMCPY
void __lsr_memcopy (
# ifdef LSR_ANSIC
	void * const dest, const void * const src, const size_t len)
# else
	dest, src, len)
	void * const dest;
	const void * const src;
	const size_t len;
# endif
{
	size_t i;
	char * const d = (char *)dest;
	const char * const s = (const char *)src;

	for ( i = 0; i < len; i++ )
	{
		d[i] = s[i];
	}
}
#endif

/* =============================================================== */

#ifndef HAVE_MEMSET
void __lsr_mem_set (
# ifdef LSR_ANSIC
	void * const dest, const char value, const size_t len)
# else
	dest, value, len)
	void * const dest;
	const char value;
	const size_t len;
# endif
{
	size_t i;
	for ( i = 0; i < len; i++ )
	{
		((char *)dest)[i] = value;
	}
}
#endif

/* =============================================================== */

#if ((defined HAVE_DLSYM) || (defined HAVE_LIBDL_DLSYM))		\
	&& (!defined HAVE_DLVSYM) && (!defined HAVE_LIBDL_DLVSYM)	\
	|| (defined __GLIBC__ && (__GLIBC__ < 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ < 1)))
# define LSR_CANT_USE_VERSIONED_FOPEN 1
/*# warning Versioned fopen is unavailable, so LibSecRm may crash on some glibc versions.*/
#else
# undef LSR_CANT_USE_VERSIONED_FOPEN
#endif

/* =============================================================== */

int LSR_ATTR ((constructor))
__lsr_main (LSR_VOID)
{
	if ( __lsr_is_initialized == LSR_INIT_STAGE_NOT_INITIALIZED )
	{
		__lsr_set_internal_function (1);
		/* Get pointers to the original functions: */

		*(void **) (&__lsr_real_unlink)			= dlsym  (RTLD_NEXT, "unlink");
		*(void **) (&__lsr_real_remove)			= dlsym  (RTLD_NEXT, "remove");
		*(void **) (&__lsr_real_unlinkat)		= dlsym  (RTLD_NEXT, "unlinkat");
		*(void **) (&__lsr_real_rmdir)			= dlsym  (RTLD_NEXT, "rmdir");
		/* Libtrash: funny interaction fixed! when dlsym() was used instead of dlvsym(),
		   GNU libc would give us a pointer to an older version of fopen() and
		   subsequently crash if the calling code tried to use, e.g., getwc().
		   YES, THIS MUST BE 2.1 !
		   */
#ifdef LSR_CANT_USE_VERSIONED_FOPEN
		*(void **) (&__lsr_real_fopen64)		= dlsym  (RTLD_NEXT, "fopen64");
#else
		*(void **) (&__lsr_real_fopen64)		= dlvsym (RTLD_NEXT, "fopen64", "GLIBC_2.1");
		if ( __lsr_real_fopen64 == NULL )
		{
			*(void **) (&__lsr_real_fopen64)	= dlsym (RTLD_NEXT, "fopen64");
		}
#endif
		*(void **) (&__lsr_real_freopen64)		= dlsym  (RTLD_NEXT, "freopen64");
		*(void **) (&__lsr_real_open64)			= dlsym  (RTLD_NEXT, "open64");
		*(void **) (&__lsr_real_openat64)		= dlsym  (RTLD_NEXT, "openat64");

		*(void **) (&__lsr_real_truncate64)		= dlsym  (RTLD_NEXT, "truncate64");
		*(void **) (&__lsr_real_ftruncate64)		= dlsym  (RTLD_NEXT, "ftruncate64");
		*(void **) (&__lsr_real_creat64)		= dlsym  (RTLD_NEXT, "creat64");
#ifdef LSR_CANT_USE_VERSIONED_FOPEN
		*(void **) (&__lsr_real_fopen)			= dlsym  (RTLD_NEXT, "fopen");
#else
		*(void **) (&__lsr_real_fopen)			= dlvsym (RTLD_NEXT, "fopen", "GLIBC_2.1");
		if ( __lsr_real_fopen == NULL )
		{
			*(void **) (&__lsr_real_fopen)		= dlsym  (RTLD_NEXT, "fopen");
		}
#endif
		*(void **) (&__lsr_real_freopen)		= dlsym  (RTLD_NEXT, "freopen");
		*(void **) (&__lsr_real_open)			= dlsym  (RTLD_NEXT, "open");
		*(void **) (&__lsr_real_openat)			= dlsym  (RTLD_NEXT, "openat");

		*(void **) (&__lsr_real_truncate)		= dlsym  (RTLD_NEXT, "truncate");
		*(void **) (&__lsr_real_ftruncate)		= dlsym  (RTLD_NEXT, "ftruncate");
		*(void **) (&__lsr_real_creat)			= dlsym  (RTLD_NEXT, "creat");
		*(void **) (&__lsr_real_posix_fallocate)	= dlsym  (RTLD_NEXT, "posix_fallocate");
		*(void **) (&__lsr_real_posix_fallocate64)	= dlsym  (RTLD_NEXT, "posix_fallocate64");
		*(void **) (&__lsr_real_fallocate)		= dlsym  (RTLD_NEXT, "fallocate");

		/* memory-related functions: */
		*(void **) (&__lsr_real_malloc)			= dlsym  (RTLD_NEXT, "malloc");
		*(void **) (&__lsr_real_psx_memalign)		= dlsym  (RTLD_NEXT, "posix_memalign");
		*(void **) (&__lsr_real_valloc)			= dlsym  (RTLD_NEXT, "valloc");
		*(void **) (&__lsr_real_pvalloc)		= dlsym  (RTLD_NEXT, "pvalloc");
		*(void **) (&__lsr_real_memalign)		= dlsym  (RTLD_NEXT, "memalign");
		*(void **) (&__lsr_real_aligned_alloc)		= dlsym  (RTLD_NEXT, "aligned_alloc");
		*(void **) (&__lsr_real_brk)			= dlsym  (RTLD_NEXT, "brk");
		*(void **) (&__lsr_real_sbrk)			= dlsym  (RTLD_NEXT, "sbrk");


#if (!defined __STRICT_ANSI__) && (defined HAVE_SRANDOM) && (defined HAVE_RANDOM)
# if (defined HAVE_TIME_H) || (defined HAVE_SYS_TIME_H) || (defined TIME_WITH_SYS_TIME)
		srandom (0xdeafface * ((unsigned int) time (NULL) & 0x0FFFFFFFF));
# else
		srandom (0xdeafface);
# endif

#else

# if (defined HAVE_TIME_H) || (defined HAVE_SYS_TIME_H) || (defined TIME_WITH_SYS_TIME)
		__lsr_srand(0xdeafface * ((unsigned int) time (NULL) & 0x0FFFFFFFF));
		/*srand (0xdeafface*(unsigned long int) time (NULL));*/
# else
		__lsr_srand(0xdeafface);
		/*srand (0xdeafface);*/
# endif
#endif
		__lsr_set_internal_function (0);
		__lsr_is_initialized = LSR_INIT_STAGE_FULLY_INITIALIZED;
	}
	return 0;
}

/* =============================================================== */
/* Functions returning pointers to real functions, so that the variables
   can't be overwritten by user code with dlsym(): */
/* =============================================================== */

i_cp		__lsr_real_unlink_location (LSR_VOID)
{
	return __lsr_real_unlink;
}

/* =============================================================== */

i_cp		__lsr_real_remove_location (LSR_VOID)
{
	return __lsr_real_remove;
}

/* =============================================================== */

i_i_cp_i	__lsr_real_unlinkat_location (LSR_VOID)
{
	return __lsr_real_unlinkat;
}

/* =============================================================== */

i_cp		__lsr_real_rmdir_location (LSR_VOID)
{
	return __lsr_real_rmdir;
}


/* =============================================================== */

fp_cp_cp	__lsr_real_fopen64_location (LSR_VOID)
{
	return __lsr_real_fopen64;
}

/* =============================================================== */

fp_cp_cp_fp	__lsr_real_freopen64_location (LSR_VOID)
{
	return __lsr_real_freopen64;
}

/* =============================================================== */

i_cp_i_		__lsr_real_open64_location (LSR_VOID)
{
	return __lsr_real_open64;
}

/* =============================================================== */

i_i_cp_i_	__lsr_real_openat64_location (LSR_VOID)
{
	return __lsr_real_openat64;
}

/* =============================================================== */

i_cp_o64	__lsr_real_truncate64_location (LSR_VOID)
{
	return __lsr_real_truncate64;
}

/* =============================================================== */

i_i_o64		__lsr_real_ftruncate64_location (LSR_VOID)
{
	return __lsr_real_ftruncate64;
}

/* =============================================================== */

i_cp_mt		__lsr_real_creat64_location (LSR_VOID)
{
	return __lsr_real_creat64;
}


/* =============================================================== */

fp_cp_cp	__lsr_real_fopen_location (LSR_VOID)
{
	return __lsr_real_fopen;
}

/* =============================================================== */

fp_cp_cp_fp	__lsr_real_freopen_location (LSR_VOID)
{
	return __lsr_real_freopen;
}

/* =============================================================== */

i_cp_i_		__lsr_real_open_location (LSR_VOID)
{
	return __lsr_real_open;
}

/* =============================================================== */

i_i_cp_i_	__lsr_real_openat_location (LSR_VOID)
{
	return __lsr_real_openat;
}

/* =============================================================== */

i_cp_o		__lsr_real_truncate_location (LSR_VOID)
{
	return __lsr_real_truncate;
}

/* =============================================================== */

i_i_o		__lsr_real_ftruncate_location (LSR_VOID)
{
	return __lsr_real_ftruncate;
}

/* =============================================================== */

i_cp_mt		__lsr_real_creat_location (LSR_VOID)
{
	return __lsr_real_creat;
}


/* =============================================================== */

i_o_o		__lsr_real_posix_fallocate_location (LSR_VOID)
{
	return __lsr_real_posix_fallocate;
}

/* =============================================================== */

i_o64_o64	__lsr_real_posix_fallocate64_location (LSR_VOID)
{
	return __lsr_real_posix_fallocate64;
}

/* =============================================================== */

i_i_i_o_o	__lsr_real_fallocate_location (LSR_VOID)
{
	return __lsr_real_fallocate;
}

/* =============================================================== */
/* memory-related functions: */
/* =============================================================== */

f_s		__lsr_real_malloc_location (LSR_VOID)
{
	return __lsr_real_malloc;
}

/* =============================================================== */

vpp_s_s		__lsr_real_psx_memalign_location (LSR_VOID)
{
	return __lsr_real_psx_memalign;
}

/* =============================================================== */

f_s		__lsr_real_valloc_location (LSR_VOID)
{
	return __lsr_real_valloc;
}

/* =============================================================== */

f_s		__lsr_real_pvalloc_location (LSR_VOID)
{
	return __lsr_real_pvalloc;
}

/* =============================================================== */

f_s_s		__lsr_real_memalign_location (LSR_VOID)
{
	return __lsr_real_memalign;
}

/* =============================================================== */

f_s_s		__lsr_real_aligned_alloc_location (LSR_VOID)
{
	return __lsr_real_aligned_alloc;
}

/* =============================================================== */

f_vp		__lsr_real_brk_location (LSR_VOID)
{
	return __lsr_real_brk;
}

/* =============================================================== */

f_ip		__lsr_real_sbrk_location (LSR_VOID)
{
	return __lsr_real_sbrk;
}
